server/ai_mediaserver.go

```diff
@@ -64,94 +64,56 @@ func startAIMediaServer(ls *LivepeerServer) error {
 
 	openapi3filter.RegisterBodyDecoder("image/png", openapi3filter.FileBodyDecoder)
 
-	ls.HTTPMux.Handle("/text-to-image", oapiReqValidator(ls.TextToImage()))
-	ls.HTTPMux.Handle("/image-to-image", oapiReqValidator(ls.ImageToImage()))
-	ls.HTTPMux.Handle("/upscale", oapiReqValidator(ls.Upscale()))
+	ls.HTTPMux.Handle("/text-to-image", oapiReqValidator(handle(ls, jsonDecoder[worker.GenTextToImageJSONRequestBody], processTextToImage)))
+	ls.HTTPMux.Handle("/image-to-image", oapiReqValidator(handle(ls, multipartDecoder[worker.GenImageToImageMultipartRequestBody], processImageToImage)))
+	ls.HTTPMux.Handle("/upscale", oapiReqValidator(handle(ls, multipartDecoder[worker.GenUpscaleMultipartRequestBody], processUpscale)))
 	ls.HTTPMux.Handle("/image-to-video", oapiReqValidator(ls.ImageToVideo()))
 	ls.HTTPMux.Handle("/image-to-video/result", ls.ImageToVideoResult())
-	ls.HTTPMux.Handle("/audio-to-text", oapiReqValidator(ls.AudioToText()))
+	ls.HTTPMux.Handle("/audio-to-text", oapiReqValidator(handle(ls, multipartDecoder[worker.GenAudioToTextMultipartRequestBody], processAudioToText)))
 	ls.HTTPMux.Handle("/llm", oapiReqValidator(ls.LLM()))
-	ls.HTTPMux.Handle("/segment-anything-2", oapiReqValidator(ls.SegmentAnything2()))
+	ls.HTTPMux.Handle("/segment-anything-2", oapiReqValidator(handle(ls, multipartDecoder[worker.GenSegmentAnything2MultipartRequestBody], processSegmentAnything2)))
 
 	return nil
 }
 
-func (ls *LivepeerServer) TextToImage() http.Handler {
+// Decoder for JSON requests
+func jsonDecoder[T any](req *T, r *http.Request) error {
+	return json.NewDecoder(r.Body).Decode(req)
+}
+
+// Decoder for Multipart requests
+func multipartDecoder[T any](req *T, r *http.Request) error {
+	multiRdr, err := r.MultipartReader()
+	if err != nil {
+		return err
+	}
+	return runtime.BindMultipart(req, *multiRdr)
+}
+
+func handle[I, O any](ls *LivepeerServer, decoderFunc func(*I, *http.Request) error, processorFunc func(context.Context, aiRequestParams, I) (O, error)) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		remoteAddr := getRemoteAddr(r)
 		ctx := clog.AddVal(r.Context(), clog.ClientIP, remoteAddr)
 		requestID := string(core.RandomManifestID())
 		ctx = clog.AddVal(ctx, "request_id", requestID)
 
-		var req worker.GenTextToImageJSONRequestBody
-		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
-			respondJsonError(ctx, w, err, http.StatusBadRequest)
-			return
-		}
-
-		clog.V(common.VERBOSE).Infof(ctx, "Received TextToImage request prompt=%v model_id=%v", req.Prompt, *req.ModelId)
+		//clog.V(common.VERBOSE).Infof(ctx, "Received TextToImage request prompt=%v model_id=%v", req.Prompt, *req.ModelId)
+		clog.V(common.VERBOSE).Infof(ctx, "Received TextToImage request")
```

Comment:
Should this be more generic now? Are we losing debug info?

Reply:
I've shifted the log lines further down into the processing part so that we can keep this code as simple as possible, I would have had to do type assertions etc. https://github.com/livepeer/go-livepeer/pull/3203/commits/35a2d34b84df27d49b05bf2ae9c2b853bb5843b1